src/Datadog.Trace/RuntimeMetrics/PerformanceCountersListener.cs

```diff
@@ -132,50 +133,100 @@ private void TryUpdateCounter(string path, PerformanceCounterWrapper counter, re
             lastValue = value;
         }
 
-        private Tuple<string, bool> GetInstanceName()
+        private InstanceName GetInstanceName()
         {
-            var instanceNames = _memoryCategory.GetInstanceNames().Where(n => n.StartsWith(_processName)).ToArray();
+            var allInstances = _memoryCategory.GetInstanceNames();
+            var fullProcessName = $"{_processName}_p{_processId}_r";
 
-            var fullName = instanceNames.FirstOrDefault(n => n.StartsWith($"{_processName}_p{_processId}_r"));
-
-            if (fullName != null)
+            string tmpName = null;
+            foreach (var insName in allInstances)
             {
-                return Tuple.Create(fullName, true);
-            }
+                if (insName.StartsWith(fullProcessName))
+                {
+                    // If we have a fullname match we return that
+                    return new InstanceName(insName, true);
+                }
 
-            if (instanceNames.Length == 1)
-            {
-                return Tuple.Create(instanceNames[0], false);
+                if (insName.StartsWith(_processName))
+                {
+                    if (tmpName is null)
+                    {
+                        // if is the first relaxed match we store it
+                        tmpName = insName;
+                    }
+                    else
+                    {
+                        // If we have more than 1 relaxed match we call the GetSimpleInstanceName()
+                        return new InstanceName(GetSimpleInstanceName(), false);
+                    }
+                }
             }
 
-            return Tuple.Create(GetSimpleInstanceName(), false);
+            // if we are here then or we didn't found a relaxed match or we have only 1.
+            return new InstanceName(tmpName, false);
         }
 
         private string GetSimpleInstanceName()
         {
-            var instanceNames = _memoryCategory.GetInstanceNames().Where(n => n.StartsWith(_processName)).ToArray();
-
-            if (instanceNames.Length == 1)
-            {
-                return instanceNames[0];
-            }
-
-            foreach (var name in instanceNames)
+            var allInstanceNames = _memoryCategory.GetInstanceNames();
```

Comment:
I have mixed feelings about making this method so complex even though it's called once every 10 seconds, but your call